packages/atlas-service/src/make-automation-agent-op-request.ts

```diff
@@ -0,0 +1,186 @@
+    name: string;
+  };
+};
+
+type AutomationAgentRequestOpTypes = keyof AutomationAgentRequestTypes;
+
+type AutomationAgentRequestResponse<
+  OpType extends AutomationAgentRequestOpTypes
+> = {
+  _id: string;
+  requestType: OpType;
+};
+
+function assertAutomationAgentRequestResponse<
+  OpType extends AutomationAgentRequestOpTypes
+>(
+  json: any,
+  opType: OpType
+): asserts json is AutomationAgentRequestResponse<OpType> {
+  if (
+    Object.prototype.hasOwnProperty.call(json, '_id') &&
+    Object.prototype.hasOwnProperty.call(json, 'requestType') &&
+    json.requestType === opType
+  ) {
+    return;
+  }
+  throw new Error(
+    'Got unexpected backend response for automation agent request'
+  );
+}
+
+export type AutomationAgentAwaitResponseTypes = {
+  listIndexStats: {
+    collName: string;
+    dbName: string;
+    indexName: string;
+    indexProperties: { label: string; properties: Record<string, unknown> }[];
+    indexType: { label: string };
+    keys: { name: string; value: string | number };
+    sizeBytes: number;
+    status: 'rolling build' | 'building' | 'exists';
+  }[];
+  dropIndex: never[];
+};
+
+type AutomationAgentAwaitOpTypes = keyof AutomationAgentAwaitResponseTypes;
+
+type AutomationAgentAwaitResponse<OpType extends AutomationAgentAwaitOpTypes> =
+  {
+    _id: string;
+    requestID: string;
+    requestType: OpType;
+    response: AutomationAgentAwaitResponseTypes[OpType];
+    type: OpType;
+  };
+
+function assertAutomationAgentAwaitResponse<
+  OpType extends AutomationAgentAwaitOpTypes
+>(
+  json: any,
+  opType: OpType
+): asserts json is AutomationAgentAwaitResponse<OpType> {
+  if (
+    Object.prototype.hasOwnProperty.call(json, '_id') &&
+    Object.prototype.hasOwnProperty.call(json, 'requestType') &&
+    Object.prototype.hasOwnProperty.call(json, 'response') &&
+    json.requestType === opType
+  ) {
+    return;
+  }
+  throw new Error(
+    'Got unexpected backend response for automation agent request await'
+  );
+}
+
+type PickAwaitResponse<OpType extends AutomationAgentAwaitOpTypes> =
+  AutomationAgentAwaitResponse<OpType>['response'];
+
+/**
+ * Helper type that maps whatever is returned by automation agent in response
+ * prop as follows:
+ *
+ * empty array -> undefined
+ * array with one item -> unwrapped item
+ * array of items -> array of items
+ */
+export type UnwrappedAutomationAgentAwaitResponse<
+  OpType extends AutomationAgentAwaitOpTypes
+> = PickAwaitResponse<OpType> extends never[]
+  ? undefined
+  : PickAwaitResponse<OpType> extends [infer UnwrappedResponse]
+  ? UnwrappedResponse
+  : PickAwaitResponse<OpType> extends Array<unknown>
+  ? PickAwaitResponse<OpType>
+  : never;
+
+function unwrapAutomationAgentAwaitResponse(
+  json: any,
+  opType: 'listIndexStats'
+): UnwrappedAutomationAgentAwaitResponse<'listIndexStats'>;
+function unwrapAutomationAgentAwaitResponse(
+  json: any,
+  opType: 'dropIndex'
+): UnwrappedAutomationAgentAwaitResponse<'dropIndex'>;
```

Comment:
A thing I didn't know about typescript before is that type narrowing doesn't apply to generics the same way it applies to typed values, so I had to be very explicit here for the unwrapping types and provide function overloads